### PR TITLE
Favorite Opponent and Rival Methods

### PR DESCRIPTION
**Pull request by stellakunzang**
Added favorite opponent and rival methods and tests, and helper methods and tests, to both GameTeam and StatTracker classes.

Also added winningest_coach and worst_coach methods (with helper methods), which were refactored to eliminate some excessively nested logic. All have tests.

### CODE REVIEW

File: lib/game_team.rb [hunk 1]
```diff
@@ -20,29 +20,70 @@ def self.from_csv(csv_file_path)
...
+    visitor_wins = (all.find_all {|gt| gt.hoa == "home" && gt.result == "LOSS" }).count.to_f
     ((visitor_wins / self.home_games) * 100).round(2)
   end
 
   def self.percentage_ties
-    games_count = @@all.count.to_f
-    ties_count = (@@all.find_all { |gt| gt.result == "TIE"}).count.to_f
+    games_count = all.count.to_f
+    ties_count = (all.find_all { |gt| gt.result == "TIE"}).count.to_f
     ((ties_count / games_count) * 100).round(2)
```
Comment (stellakunzang) [inline lib/game_team.rb:39]: the transition from `@@all` to `all` method was completely painless!

File: lib/game_team.rb [hunk 2]
```diff
@@ -20,29 +20,70 @@ def self.from_csv(csv_file_path)
...
+    coaches_in_season(season_id).each do |coach|
+      all.find_all do |gt|
+        if gt.game_id.to_s[0..3] == search_term && coach == gt.head_coach
+            results_by_coach_by_season[coach] << gt.result
+        end
+      end
+    end
+    results_by_coach_by_season
+  end
+
```
Comment (stellakunzang) [inline lib/game_team.rb:62]: Believe it or not these were actually refactored to have less ends at the end!

File: lib/game_team.rb [hunk 3]
```diff
@@ -146,6 +186,44 @@ def self.least_goals_scored(team_id)
...
 
+  def self.opponents_records(team_id)
+    game_ids = all.map {|gt| gt.game_id if gt.team_id == team_id }.compact
+    opponents_records = Hash.new { |hash, key| hash[key] = [] }
+    game_ids.each do |game_id|
+      all.find_all do |gt|
+        opponents_records[gt.team_id] << gt.result if gt.game_id == game_id && gt.team_id != team_id
+      end
+    end
+    opponents_records
```
Comment (stellakunzang) [inline lib/game_team.rb:197]: This leaves me with a hash with the team_id as the key and an array as the value that contains their results(["WIN", "LOSS", "TIE"])

File: lib/stat_tracker.rb [hunk 4]
```diff
@@ -121,5 +120,18 @@ def most_goals_scored(team_id)
   def least_goals_scored(team_id)
     GameTeam.least_goals_scored(team_id)
   end
+  
+  def favorite_opponent(team_id)
+    team = Team.all.find do |team|
+      team.team_id == GameTeam.favorite_opponent_id(team_id)
+    end
+    team.team_name
+  end
```
Comment (stellakunzang) [inline lib/stat_tracker.rb:129]: it was pretty fun using a method from one class to get information from another class

File: test/game_team_test.rb [hunk 5]
```diff
@@ -164,6 +182,40 @@ def test_it_can_return_worst_offense_team_number
...
+    game_team3 = mock
+    game_team4 = mock
+    game_team1.stubs(:team_id).returns(1)
+    game_team2.stubs(:team_id).returns(2)
+    game_team3.stubs(:team_id).returns(3)
+    game_team4.stubs(:team_id).returns(4)
+    GameTeam.stubs(:opponents_records).returns({2=>["LOSS", "WIN", "LOSS","LOSS"], 3=>["WIN", "WIN", "WIN"], 4=>["LOSS", "WIN"]})
+    assert_equal 2, GameTeam.favorite_opponent_id(1)
+  end
+
```
Comment (stellakunzang) [inline test/game_team_test.rb:205]: I used mocks and stubs! Hooray!

File: test/game_test.rb [hunk 6]
```diff
@@ -226,6 +226,12 @@ def test_grouped_season_returns_array_of_games_grouped_by_season
     assert_equal 20, Game.all.count
   end
 
+# the code was changed and this test is no longer passing 
+  def test_it_returns_all_games_by_seasons_given_team_id
+    assert_equal ({20122013 => 2, 20142015 => 4}), Game.games_by_season(3)
+    assert_equal ({20122013 => 2}), Game.games_by_season(6)
+  end
+
```
Comment (stellakunzang) [inline test/game_test.rb:234]: I'm not sure who wrote this method but something changed and now the test doesn't pass. Maybe I didn't fix the merge conflicts correct, but it seems important that before doing a pull request we should run all our tests and make sure everything is passing one last time. I wasn't sure how to fix it, so I left it with a comment.

File: lib/game_team.rb (hunk 1 above)
Comment (perryr16) [reply lib/game_team.rb:39]: image

File: lib/game_team.rb (hunk 2 above)
Comment (perryr16) [reply lib/game_team.rb:62]: progress!

File: test/game_test.rb (hunk 6 above)
Comment (perryr16) [reply test/game_test.rb:234]: I'll take a look. thanks for the comment

**Review by perryr16 — COMMENTED** [review]
functional code, I'll look at that error

File: lib/game_team.rb (hunk 2 above)
Comment (Rostammahabadi) [reply lib/game_team.rb:62]: We can refactor this to utilize find_by(season_id) which would delete out a few lines

File: lib/game_team.rb (hunk 3 above)
Comment (Rostammahabadi) [reply lib/game_team.rb:197]: This would be a great candidate for a nested hash

**Pull request merged**